test/mpi/dtpools/src/dtpools_internal.c

```diff
@@ -0,0 +1,1167 @@
+    unsigned char *uchar_ptr;
+    unsigned short int *ushort_ptr;
+    unsigned int *uint_ptr;
+    unsigned long int *ulong_ptr;
+    unsigned long long int *ulong_long_ptr;
+    float *float_ptr;
+    double *double_ptr;
+    long double *long_double_ptr;
+    int8_t *int8_ptr;
+    int16_t *int16_ptr;
+    int32_t *int32_ptr;
+    int64_t *int64_ptr;
+    uint8_t *uint8_ptr;
+    uint16_t *uint16_ptr;
+    uint32_t *uint32_ptr;
+    uint64_t *uint64_ptr;
+};
+
+/* --------------------------------------------------------- */
+/* Utility Functions                                         */
+/* --------------------------------------------------------- */
+
+void DTPI_Print_error(int errcode) {
+    int errclass, slen;
+    char string[ERR_STRING_MAX_LEN];
+
+    MPI_Error_class(errcode, &errclass);
+    MPI_Error_string(errcode, string, &slen);
+
+    fprintf(stdout, "Error class %d (%s)\n", errclass, string);
+    fflush(stdout);
+}
+
+void DTPI_Init_creators(DTPI_Creator * creators)
+{
+    memset(creators, 0, sizeof(*creators));
+    creators[DTPI_OBJ_LAYOUT_SIMPLE__BASIC] = DTPI_Basic_create;
+    creators[DTPI_OBJ_LAYOUT_SIMPLE__CONTIG] = DTPI_Contig_create;
+    creators[DTPI_OBJ_LAYOUT_SIMPLE__VECTOR] = DTPI_Vector_create;
+    creators[DTPI_OBJ_LAYOUT_SIMPLE__INDEXED] = DTPI_Indexed_create;
+    creators[DTPI_OBJ_LAYOUT_SIMPLE__BLOCK_INDEXED] = DTPI_Block_indexed_create;
+    creators[DTPI_OBJ_LAYOUT_LARGE_BLK__VECTOR] = DTPI_Vector_create;
+    creators[DTPI_OBJ_LAYOUT_LARGE_BLK__INDEXED] = DTPI_Indexed_create;
+    creators[DTPI_OBJ_LAYOUT_LARGE_BLK__BLOCK_INDEXED] = DTPI_Block_indexed_create;
+    creators[DTPI_OBJ_LAYOUT_LARGE_CNT__VECTOR] = DTPI_Vector_create;
+    creators[DTPI_OBJ_LAYOUT_LARGE_CNT__INDEXED] = DTPI_Indexed_create;
+    creators[DTPI_OBJ_LAYOUT_LARGE_CNT__BLOCK_INDEXED] = DTPI_Block_indexed_create;
+    creators[DTPI_OBJ_LAYOUT_LARGE_BLK_STRD__VECTOR] = DTPI_Vector_create;
+    creators[DTPI_OBJ_LAYOUT_LARGE_BLK_STRD__INDEXED] = DTPI_Indexed_create;
+    creators[DTPI_OBJ_LAYOUT_LARGE_BLK_STRD__BLOCK_INDEXED] = DTPI_Block_indexed_create;
+    creators[DTPI_OBJ_LAYOUT_LARGE_CNT_STRD__VECTOR] = DTPI_Vector_create;
+    creators[DTPI_OBJ_LAYOUT_LARGE_CNT_STRD__INDEXED] = DTPI_Indexed_create;
+    creators[DTPI_OBJ_LAYOUT_LARGE_CNT_STRD__BLOCK_INDEXED] = DTPI_Block_indexed_create;
+}
+
+static void DTPI_Type_init_buf(struct params *par, const char *basic_type_name, void *buf)
+{
+    int i, j, k;
+    int start, stride, count;
+    MPI_Aint type_count, type_stride, type_blklen, type_offset;
+    char *buf_ptr;
+    struct cast_ptrs ptrs;
+
+    if (par->count <= 0) {
+        return;
+    }
+
+    start = par->start;
+    count = par->count;
+    stride = par->stride;
+    type_count = par->type_count;
+    type_stride = par->type_stride;
+    type_blklen = par->type_blklen;
+    type_offset = par->type_offset;
+
+    buf_ptr = (char *) buf + type_offset;
+    if (!STRCMP(basic_type_name, "MPI_CHAR")) {
```

Comment:
Why do you use string `basic_type_name` for switching between types ? Comparing `MPI_Datatype` should be much faster.